#!/bin/sh
#Align reads pre-digested with DpnII and HinfI using parasplit.
#If not using parasplit, one might use the "cutsite" mode of hicstuff. In this case replace the "mode" argument by "cutsite", and in the hicstuff pipeline command replace the "-m normal" argument by "-m $mode"
genome="S288c_DSB_chr3_rDNA" #reference genome
thread=6 #number of threads
mode="parasplit" #see comment above
quality=20 #alignment quality
hicstuffversion="v324"
fastqdir='/mnt/e/Science/Reads/'
outputdir='/mnt/f/outputdir/DR07_UMR5239_Experiments/HiC/'
SAMP=("CD23" "CD24" "CD25" "CD26" "CD27" "CD28") #list of sample to run, with suffix in the form .end1/2.fastq.gz

for sample in "${SAMP[@]}" ; do
	output=$sample'_'$genome'_'$mode'_q'$quality'_'$hicstuffversion
	
	echo $sample "started"
	#pre-digest the reads with parasplit
	parasplit -sf $fastqdir$sample'.end1.fastq.gz' -sr $fastqdir$sample'.end2.fastq.gz' -of $fastqdir$sample'_parasplit_digest_for.fq.gz' -or $fastqdir$sample'_parasplit_digest_rev.fq.gz' -le DpnII,HinfI -nt $thread -m all

	mkdir $sample;
	#align reads on ref genome and generate sparse matrices in fragments or binned in graal (3-columns) format
	hicstuff pipeline -m normal -M graal -t $thread -e DpnII,HinfI -q $quality -g 'Genomes/'$genome'/'$genome -n -p -o $sample -d -f -F -D $fastqdir$sample'_parasplit_digest_for.fq.gz' $fastqdir$sample'_parasplit_digest_rev.fq.gz'
	#rebin the matrix at 1kb in graal format
	hicstuff rebin -b 1kb -f $sample'/fragments_list.txt' -c $sample'/info_contigs.txt' $sample'/abs_fragments_contacts_weighted.txt' $sample'/'$output'_1kb'
	#convert the binned graal matrix into cooler format, ICE-balance it, and zoomify it at 1, 2, 5, 10, 15, and 20kb
	mkdir $sample'/Cool';
	hicstuff convert -f $sample'/'$output'_1kb.frags.tsv' -c $sample'/'$output'_1kb.chr.tsv' $sample'/'$output'_1kb.mat.tsv' $sample'/Cool/'$output'_1kb'
	cooler balance -p $thread $sample'/Cool/'$output'_1kb.cool'
	cooler zoomify \
            --resolutions 1000,2000,5000,10000,15000,20000 \
            --balance \
            -n $thread \
            -o $sample'/Cool/'$output'.mcool' \
            $sample'/Cool/'$output'_1kb.cool'
done

#compute bam files and binned coverage from alignement files generated by hicstuff -> check expected presence of mutations in the strain, coverage around DSB site, and abnormal CNV/aneuploidies
for sample in "${SAMP[@]}" ; do  	
	output=$sample'_'$genome'_'$mode'_q'$quality'_'$hicstuffversion
	mkdir $sample'/Coverage'
	samtools sort -l 5 -o $sample'/tmp/for_sorted.bam' -@ $thread $sample'/tmp/for.bam'
	samtools sort -l 5 -o $sample'/tmp/rev_sorted.bam' -@ $thread $sample'/tmp/rev.bam'
	samtools merge -@ $thread $sample'/tmp/pairs_sorted.bam' $sample'/tmp/for_sorted.bam' $sample'/tmp/rev_sorted.bam'
	samtools index -b $sample'/tmp/pairs_sorted.bam'
	#bin the bam file into a bedgraph for easy visualization in IGV using tinycov. Specify the binning -r in bp and the shift of sliding window in bp.
	tinycov covplot -r 500 -s 500 -n $output'_coverage_r500_s500' -t $sample'/Coverage/'$output'_coverage_r500_s500.bedgraph' -o $sample'/Coverage/'$output'_coverage_r500_s500.png' -p 0 $sample'/tmp/pairs_sorted.bam'
done
